src/main.rs

```diff
@@ -10,30 +11,212 @@ mod state;
+
+enum EventType {
+  IssueComment,
+  PullRequestReviewComment,
+  PullRequest,
+  Push,
+  UnknownEvent,
+}
+
+struct Deserialize;
+
+struct EventInfo;
+impl typemap::Key for EventInfo { type Value = EventType; }
+
+struct IsIssueComment;
+impl typemap::Key for IsIssueComment { type Value = IssueCommentEvent; }
+
+struct IsPullRequestReviewComment;
+impl typemap::Key for IsPullRequestReviewComment { type Value = PullRequestReviewCommentEvent; }
+
+struct IsPullRequest;
+impl typemap::Key for IsPullRequest { type Value = PullRequestEvent; }
+
+struct IsPush;
+impl typemap::Key for IsPush { type Value = PushEvent; }
+
+
+impl BeforeMiddleware for Deserialize {
+  fn before(&self, req: &mut Request) -> IronResult<()> {
+    let mut payload = String::new();
+    req.body.read_to_string(&mut payload).unwrap();
+
+    let headers = req.headers.clone();
+    let event_header = headers.get::<GithubEventHeader>();
+    event_header.map(|header| {
+      match header.event_name.as_ref() {
+        "issue_comment" => {
+          req.extensions.insert::<EventInfo>(EventType::IssueComment);
+          req.extensions.insert::<IsIssueComment>(json::decode(&payload).unwrap());
+        },
+        "pull_request_review_comment" => {
+          req.extensions.insert::<EventInfo>(EventType::PullRequestReviewComment);
+          req.extensions.insert::<IsPullRequestReviewComment>(json::decode(&payload).unwrap());
+        },
+        "pull_request" => {
+          req.extensions.insert::<EventInfo>(EventType::PullRequest);
+          req.extensions.insert::<IsPullRequest>(json::decode(&payload).unwrap());
+        },
+        "push" => {
+          req.extensions.insert::<EventInfo>(EventType::Push);
+          req.extensions.insert::<IsPush>(json::decode(&payload).unwrap());
+        },
+        _ => {req.extensions.insert::<EventInfo>(EventType::UnknownEvent);}
+      }
+    });
+    Ok(())
+  }
+}
+
 
 fn handle_webhooks(req: &mut Request) -> IronResult<Response> {
-  println!("webhook hit {:?}", req);
-  Ok(Response::with((status::Ok, "")))
+  let possible_event_type = req.extensions.get::<EventInfo>();
+
+  match possible_event_type {
+    Some(&EventType::IssueComment) => Ok(Response::with((status::Accepted, "{\"body\":\"Successful recv of issue comment\"}"))),
+    Some(&EventType::PullRequestReviewComment) => Ok(Response::with((status::Accepted, "{\"body\":\"Successful recv of pull request review comment\"}"))),
+    Some(&EventType::PullRequest) => Ok(Response::with((status::Accepted, "{\"body\":\"Successful recv of pull request\"}"))),
+    Some(&EventType::Push) => Ok(Response::with((status::Accepted, "{\"body\":\"Successful recv of push\"}"))),
+    Some(&EventType::UnknownEvent) => Ok(Response::with((status::Accepted, "{\"body\":\"Recv an unhandled event\"}"))),
+    None => Ok(Response::with((status::Accepted, "{\"body\":\"No event header provided\"}")))
+  }
+}
+
+struct DeliverActionables {
+  issue_comment_tx: Mutex<Sender<IssueCommentEvent>>,
```

Comment:
@seanstrom @rschifflin so the issue with Sender not being `Sync` like i was talking about before is greatly alleviated by wrapping the whole thing in a mutex. Just make the request threads acquire a lock on the shared `Sender` and you're good to go.